Sources/SwiftParser/Attributes.swift

```diff
@@ -869,6 +876,152 @@ extension Parser {
+    return RawAttributeSyntax(
+      unexpectedBeforeAtSign,
+      atSignToken: atSign,
+      unexpectedBeforePackageToken,
+      attributeName: RawTypeSyntax(RawSimpleTypeIdentifierSyntax(name: packageToken, genericArgumentClause: nil, arena: self.arena)),
+      unexpectedBeforeLeftParen,
+      leftParen: leftParen,
+      argument: .packageAttributeArguments(arguments),
+      unexpectedBeforeRightParen,
+      rightParen: rightParen,
+      arena: self.arena
+    )
+  }
+
+  enum PackageLocationLabel: RawTokenKindSubset {
+    case id
+    case path
+    case url
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch lexeme {
+      case RawTokenKindMatch(.id): self = .id
+      case RawTokenKindMatch(.path): self = .path
+      case RawTokenKindMatch(.url): self = .url
+      default: return nil
+      }
+    }
+
+    var rawTokenKind: RawTokenKind {
+      switch self {
+      case .id: return .keyword(.id)
+      case .path: return .keyword(.path)
+      case .url: return .keyword(.url)
+      }
+    }
+  }
+
+  enum PackageRequirementLabel: RawTokenKindSubset {
+    case branch
+    case from
+    case revision
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch lexeme {
+      case RawTokenKindMatch(.branch): self = .branch
+      case RawTokenKindMatch(.from): self = .from
+      case RawTokenKindMatch(.revision): self = .revision
+      default: return nil
+      }
+    }
+
+    var rawTokenKind: RawTokenKind {
+      switch self {
+      case .branch: return .keyword(.branch)
+      case .from: return .keyword(.from)
+      case .revision: return .keyword(.revision)
+      }
+    }
+  }
+
+  mutating func parsePackageAttributeArguments() -> RawPackageAttributeArgumentsSyntax {
+    // Parsing package location.
+    let locationLabel = self.consume(ifAnyIn: PackageLocationLabel.self) ?? missingToken(.identifier)
+    let (unexpectedBeforeLocationColon, locationColon) = self.expect(.colon)
+    let (unexpectedBeforeLocation, location) = self.expect(.stringLiteral)
+    // Parsing package requirement.
+    let (unexpectedBeforeLocReqComma, locReqComma): (RawUnexpectedNodesSyntax?, RawTokenSyntax?)
+    let (unexpectedBeforeRequirementLabel, requirementLabel): (RawUnexpectedNodesSyntax?, RawTokenSyntax?)
+    let (unexpectedBeforeRequirementColon, requirementColon): (RawUnexpectedNodesSyntax?, RawTokenSyntax?)
+    let requirement: RawExprSyntax?
+    if locationLabel.tokenKind != .keyword(.path) {
+      (unexpectedBeforeLocReqComma, locReqComma) = self.expect(.comma)
+      if let label = self.consume(ifAnyIn: PackageRequirementLabel.self) {
+        (unexpectedBeforeRequirementLabel, requirementLabel) = (nil, label)
+        (unexpectedBeforeRequirementColon, requirementColon) = self.expect(.colon)
+      } else {
+        (unexpectedBeforeRequirementLabel, requirementLabel) = (nil, nil)
+        (unexpectedBeforeRequirementColon, requirementColon) = (nil, nil)
+      }
+      requirement = self.parseExpression()
```

Comment:
Shouldn’t this also be in the `if let label` branch?

---

Open question: 

Do we want to allow arbitrary expressions here? If we only want to allow e.g. `"1.1.0"` or `"1.1.0"..<"1.2.0"` it might make sense to create a new syntax node for that kind of version requirement.

Reply:
Well, the current proposal handles requirement like this:
* If the given location is a `path`, there should be no requirement; otherwise we always have one.
* If the label is `from`, it accepts a semantic version in string literal;
* If the label is `branch`, it accepts a branch name in string literal;
* If the label is `revision`, it accepts a commit hash in string literal.
* If there's no label, it accepts either a range expression (eg. `"1.0.0"..<"1.1.0"`) or a member access expression (eg. `.upToNextMinor(from: "1.0.0")`).

I haven't decided if we want to simplify the interface (making it less like `PackageDescription`), so we're now accepting arbitrary expressions here.

Reply:
Since this is only a draft, there're plenty of rooms for refactors, like:
* Extracting the logic of `@_package` into a separate file like `@available` does;
* Implementing fully semantical parsing logic including expression diagnostic and even semantics versioning;
* Consuming all parameters at a whole (just like a tuple) and implementing semantical error handling.

WDYT?

Reply:
> * Extracting the logic of `@_package` into a separate file like `@available` does;

I don’t think starting a new file makes sense for ~100 LOC. If things grow further, 

> * Implementing fully semantical parsing logic including expression diagnostic and even semantics versioning

What do you mean by “fully semantic parsing logic including expression diagnostics”. 

> * Introducing additional node types for package attribute;

There’s always a trade-off here between expressibility and trying to keep the syntax tree simple. The main consideration would be if there are invariants in the syntax tree that clients might care about but which are not expressed in the syntax tree at the moment (e.g. if two options are mutually exclusive). I don’t see anything like that at the moment, so maybe it’s fine as-is.

> * Consuming all parameters at a whole (just like a tuple) and implementing semantical error handling.
My gut f

Reply:
Now I've refactored the support and it grows to ~120 loc. Parser implementation should look better with new nodes added.

After checking `PackageDescription` source, I found that wildcard requirement was deprecated in SwiftPM 5.6, so for a new feature it makes sense to drop the support. This means although the node is a plain `Expr` at the time, we can further assert it to be a range expression ( `semantic-version-low ( '..<' | '...' ) semantic-version-high`).

Reply:
If we just allow `semantic-version-low ( '..<' | '...' ) semantic-version-high`, I think it would make sense to represent the version requirement in the syntax tree as three nodes `VersionTuple`, `BinaryOperator` and `VersionTuple`.

Reply:
`VersionTuple` doesn't fully mimic a semantic version (which is rather complex and better handed over to other tools). I’m going to use `StringLiteralExpr` instead. Thanks for your suggestions on `BinaryOperator`, but I didn’t have idea on how to diagnose wrong operator here🤔 Should that be handled elsewhere?

Reply:
Handling the semantic version somewhere else sounds good to me. The binary operator could be anything. There isn’t any way to restrict the kinds of operators we allow in the SwiftSyntax tree.